### ISSUE TITLE
Generalize Dead-insert-elimination

### ISSUE THREAD

**Issue by s-perron** [KhronosGroup/SPIRV-Tools#1511]
Dead insert elimination looks for chains of instructions starting with an OpCompositeExtract followed by a series of OpCompositeInserts.  It then tried to remove insert whose value is never used because it is not extracted, or it replaced by another insert instruction.

This is useful but it is limited by the fact that there cannot be any instruction between the inserts.

We want to extend this to be able to 
- [x] handle instructions that treat vectors on a per component basis.   
- [x] handle intervening vector shuffle instructions,
- [x] be able to replace unused elements of OpCompositeConstruct
- [ ] handle inserts and extract for other composite types.

Another issue is that dead insert elimination repeats work.  With the longer chain of instructions that the above changes will cause, it could negatively effect compile time.

To be able to avoid special casing a lot of stuff in the current implementation, we believe it is best to create a new one where we keep track of which elements of each vector is live.  The current implementation keeps track of just the insert instructions that are needed.  This change is intrusive enough to warrant a new implementation.

**Comment by s-perron** [comment]
Sorry forgot that we need to handle structs before we can remove dead-insert-elimination.  I've added that to the description.

**Comment by s-perron** [comment]
Done what we are doing for now.
